### ISSUE TITLE
Unify QVR collection expressions with QIEC: typed map/traverse/fold, exhaustive analysis, and honest diagnostics

### ISSUE THREAD

**Issue by aaronstevenwhite** [FACTSlab/quivers#64]
## Problem

QVR has one parsed deterministic-expression surface, but it currently has three different meanings depending on which path consumes it:

1. the eager/legacy program compiler;
2. program-to-QIEC elaboration and the QIEC reference runtime; and
3. QIEC-backed transpilation.

The immediate symptom is that `map` parses and the eager compiler implements it, but QIEC cannot lower it. This is not only a missing builtin. It prevents compact finite-model code such as mapping an RSA speaker over a finite lexicon, and it exposes several related inconsistencies in lambdas, collection operations, builtins, shape semantics, expression analysis, and diagnostics.

The current behavior can report success for a program that neither `qvr run` nor `qvr transpile` can use:

```qvr
object Input : FinSet 1
object Output : Real 3

program map_gap : Input -> Output
    let xs = [1.0, 2.0, 3.0]
    let ys = map(xs, x -> x * 2.0)
    return ys

export map_gap
```

At `814b57261e55a9a893aaa963d18dc81dc5d01f80`:

```text
$ qvr check --json map-gap.qvr
{"diagnostics": [], "ok": true, ...}

$ qvr run map-gap.qvr map_gap --json
unknown entry point 'map_gap'; available: (none)

$ qvr transpile map-gap.qvr --to pyro
a `let` calls `map(...)`, which names no builtin ...
```

For a `define`, the corresponding expression fails earlier because an inline lambda is not a QIEC value:

```qvr
define twice(xs : Tensor[Real]([3])) : Tensor[Real]([3]) !{} =
    return map(xs, x -> x * 2.0)
```

```text
a lambda is not a QIEC value; QIEC values are literals, locals, operators,
tuples, builtin applications, and constructor applications
```

The desired outcome is not unrestricted higher-order QIEC. It is a coherent, typed collection surface whose finite forms elaborate to the existing first-order kernel, plus explicit diagnostics for forms whose shape or effects require additional core support.

### Surface coverage

| Form | Parser / emitter | Eager program compiler | QIEC lowering | QIEC transpilation | Gap |
|---|---:|---:|---:|---:|---|
| literals, tuples, arithmetic | yes | yes | yes | yes | Numeric and broadcasting semantics differ between the eager and QIEC paths. |
| nonempty homogeneous `[a, ...]` | yes | Python `list` | `TensorValue` | yes | The same spelling denotes different host-level values; empty and heterogeneous lists have no checked QIEC type. |
| `factor i : I in body` | yes | nested host iteration + stack | `Comprehension` | yes | This is already the checked finite-map primitive, but it only accepts a pure value body. |
| `x -> body` | yes | host closure | rejected as a value; a top-level program `let` may be retained as a one-argument macro | no general value form | Inline lambdas and `define`-local lambdas fail; curried/multiargument macro application is not coherent. |
| `map(xs, f)` | yes | yes, returning a Python list | no | no | Fixed-shape tensor map should lower to `Comprehension`. |
| `length(xs)` | yes | yes, returning `float(len(xs))` | no | no | Tensor length is statically known and should be an `Int` in QIEC. |
| `fold(xs, init, f)` | yes | partial host implementation | no | no | The comment says `fn(x)(acc)`, the implementation evaluates `fn(acc)(x)`, and there is no checked fold term. |
| `logsumexp_over(xs, f)` | yes | yes | no | no | This is derived `logsumexp(map(xs, f))` for a finite tensor. |
| `filter(xs, p)` | yes | yes, returning a variable-length list | no | no | A dynamic result length cannot inhabit `Tensor[A]([n])`; this needs a packed/existential collection or a masked fixed-shape API. |
| method calls | yes | dynamic host dispatch | only the dedicated `chart.goal_weight()` case | no general form | Arbitrary host methods should not be presented as portable QIEC expressions. |
| calls to injected host functions | yes | dynamic host dispatch | rejected | rejected | This should remain nonportable unless replaced by a declared computation or primitive. |
| pure computation call inside a value map | no usable traversal form | callable-dependent | only as `let x <- f(...)` | calls themselves transpile | There is no effect-aware `traverse`/computation comprehension, which is the form needed to map an effect-free or effectful `define` over a finite index. |

### Builtin drift

The eager compiler has 86 names in `_LET_EXPR_BUILTINS`; the checked registry exposes 60 names in `PURE_BUILTINS`; only 55 overlap. There are currently 31 eager-only names:

```text
alpha_dropout, amax, amin, argmax, argmin, celu, clamp, cummax, cummin,
cumprod, dropout, flip, glu, hardshrink, hardsigmoid, hardswish, hardtanh,
layer_norm, leaky_relu, neg, norm, prelu, rms_norm, rrelu, softmin,
softshrink, std, swish, tanhshrink, threshold, var
```

and five checked-only names:

```text
int, pow, real, weight, weight_value
```

This is not merely a difference in inventory:

- the eager call path evaluates every argument and then invokes most tensor builtins with `args[0]`, silently ignoring additional authored arguments;
- `min` and `max` are unary last-axis reductions in the eager table but are both binary scalar primitives and unary reductions in QIEC;
- the eager special case for `logsumexp` shadows the table entry and has different behavior for a single tensor argument;
- eager arithmetic uses `torch.broadcast_tensors`, while QIEC permits scalar lifting and equal tensor shapes but rejects general rank broadcasting;
- eager numeric literals become floating torch tensors, while QIEC distinguishes `Int` and `Real`; and
- `dropout`, `alpha_dropout`, and randomized `rrelu` should not be admitted as pure deterministic primitives without an effect semantics.

The tables should not be manually synchronized. The source language needs one builtin registry recording overloads, purity/effects, shape action, eager implementation, QIEC primitive/term, and per-target capability/spelling.

### Expression-walker drift

Several independent recursive walkers implement subtly different subsets of `LetExprNode`:

- `program_elaboration._free_let_names` does not descend into lambdas, factors, or method calls;
- `_unknown_call`, `_chart_construct`, and `_index_names` omit different binder-bearing forms or subexpressions;
- `step_resolution._let_expr_value_names` handles lambdas and factors but omits some literal/product variants; and
- `_substitute_let` is binder-aware for lambdas/factors but does not rewrite method calls.

This makes feature support depend on which preliminary analysis happened to see a subtree. It also makes it easy for a new AST variant to parse and emit while silently escaping dependency, unknown-call, shape, or substitution analysis.

### Diagnostics drift

`Compiler._lower_qiec` catches `qiec-program-gap`, reruns lowering with `elaborate_programs=False`, and records the message in `QiecModule.gap`. `qvr check` does not report that field. `qvr check --target pyro` then analyzes the reduced module, which contains no program entry, and also reports success. This is why the reproducer passes checking but disappears from `qvr run` and fails only when transpilation is attempted directly.

## Existing machinery to reuse

The checked kernel already has the difficult fixed-shape pieces:

- `Comprehension(binder, extent, body, result_type)` with checking, substitution, canonical serialization, reference evaluation, QIEC IR, capability analysis, and Python/Julia/JavaScript/Scheme rendering;
- `Reduction` for `sum`, `mean`, `max`, `min`, `logsumexp`, and `prod`;
- `Rowwise` for `softmax`, `log_softmax`, `cumsum`, `sort`, and `normalize`; and
- `factor`, which lowers one or more finite binders to nested `Comprehension` terms and round-trips through transpilation.

For instance, the following already checks, runs to `[2, 4, 6]`, and transpiles:

```qvr
object Axis : FinSet 3

define twice(xs : Tensor[Real]([3])) : Tensor[Real]([3]) !{} =
    return factor i : Axis in xs[i] * 2.0
```

Thus pure `map` does not need a higher-order kernel value or a backend-specific implementation. Its lambda can be consumed during elaboration and lowered to the same first-order `Comprehension` used by `factor`.

### 1. Establish one exhaustive expression traversal API

Add a binder-aware visitor/rewriter for `LetExprNode` and migrate free-name collection, unknown-call detection, chart detection, index-use analysis, substitution, program-template renaming, and similar passes to it.

Requirements:

- every tagged-union variant is handled explicitly;
- lambda parameters and factor binders shadow outer names;
- rewriting preserves source locations;
- method receivers and arguments are traversed;
- factor bodies and every case are traversed;
- tests enumerate all `LetExprNode` variants so adding a variant cannot silently create a partial walker.

This should land before adding another binder-bearing form.

### 2. Define checked collection semantics

In checked QIEC source, collection combinators should operate on homogeneous tensors. The first axis is the collection axis:

```text
map    : Tensor[A]([n, ...s]) × (Tensor[A]([...s]) -> B)
       -> Tensor[B]([n, ...shape(B)])

length : Tensor[A]([n, ...s]) -> Int
```

The lambda is elaboration syntax, not a first-class runtime closure. Mapping over a vector binds one element; mapping over a matrix binds one row. Captures are ordinary lexical captures checked in the surrounding context.

Lower:

```qvr
map(xs, x -> body)
```

to a `Comprehension` whose binder `i` ranges over the statically known leading extent and whose body is `body[x := xs[i]]`. This is beta-reduction into a first-order term, so the QIEC ABI remains first-order and all current host renderers can already execute the result.

`length(xs)` lowers to the leading shape index as an `Int`. Open symbolic extents should either lower to a checked shape-index value if the kernel gains one, or receive a precise capability diagnostic; they should not become a host reflection operation.

Permit both inline lambdas and a let-bound lambda name in combinator position. Generalize macro application by peeling curried lambdas deliberately, with one documented argument order, rather than retaining the current one-argument special case.

### 3. Add finite folds without adding general closures

Define the surface contract first:

```qvr
fold(xs, init, acc -> x -> body)
```

is a strict left fold in increasing index order. This matches the eager implementation's actual `fn(acc)(x)` order and removes the contradictory comment.

Lower recognized associative folds to existing `Reduction` terms when semantics agree. Add a first-order QIEC `Fold` value for the general case, containing:

- the input tensor and its leading extent;
- accumulator and element locals;
- the initial value;
- the body; and
- the checked accumulator/result type.

`Fold` needs the same end-to-end treatment as `Comprehension`: inference, substitution, serialization, reference evaluation, QIEC IR, capability collection, and the four host-runtime renderers. It does not require a function type.

Lower `logsumexp_over(xs, x -> score)` as `logsumexp(map(xs, x -> score))`; no dedicated kernel term is needed.

### 4. Add a computation-level finite traversal

Pure `map` cannot express the motivating nested-model case because QIEC computation calls are sequenced with `<-` and are not values. Add the computation analogue of `factor`, tentatively:

```qvr
let speakers <- traverse lexicon : Lexicon in
    pragmatic_speaker(some_meanings[lexicon], all_meaning, prior, alpha)
```

and permit multiple finite binders as `factor` does. The result prepends the binder extents to the callee result shape.

The first implementation can elaborate a statically finite traversal to a deterministic chain of `Call`/`Bind` nodes followed by `TensorValue`, in lexicographic index order. This requires no new kernel construct and naturally unions the callee's effect row. If code-size growth becomes material, a checked `Traverse` computation term can be added later with the same semantics.

The checker must enforce:

- every binder has a static finite extent;
- every iteration returns the same type;
- the declared effect row admits the callee's row;
- site identities remain distinct and stable per iteration; and
- ordering is specified, since stateful or traced effects can observe it.

This is the abstraction needed for models such as lexical-uncertainty RSA: map the literal listener over utterances, map the pragmatic speaker over lexica, reduce over the lexicon prior, and retain the listener/speaker call graph instead of spelling each cell by hand.

### 5. Treat `filter` as a separate type-design problem

Do not implement checked `filter` by pretending its result is a statically shaped tensor. Choose and document one of:

1. a packed/dependent collection such as `Σ k : Nat. Tensor[A]([k])`, with a bound `k <= n`; or
2. a fixed-shape selection result carrying values plus a Boolean mask/count.

Until that type exists, QIEC should emit a specific `qiec-collection:filter:dynamic-shape` diagnostic and point users to masking. The eager Python-list behavior may remain explicitly host-only, but it should not make a program appear QIEC-runnable or portable.

### 6. Consolidate builtin definitions and shape behavior

Replace `_LET_EXPR_BUILTINS`/`PURE_BUILTINS` drift with a declarative registry. Each entry should state:

- surface name and typed overloads;
- scalar lift, last-axis reduction, rowwise transform, or explicit shape rule;
- purity/effect classification;
- eager implementation;
- QIEC primitive or term constructor; and
- target support/spelling.

Generate the eager dispatch and QIEC lookup tables from this registry. Enforce arity before execution. Remove the `args[0]` fallback. Keep random training operations out of the pure registry unless represented by a declared effect.

For broadcasting, either retain the checked rule (scalar lift plus equal tensor shapes) and make the eager path conform, or add an explicit checked `Broadcast`/`Expand` term with shape constraints. Do not silently inherit unrestricted torch broadcasting in only one execution path.

### 7. Make gaps visible at the first check

When `Compiler._lower_qiec` records `QiecModule.gap`, all user-facing tools must surface it:

- `qvr check` reports a diagnostic rather than `OK` for an exported program that was omitted;
- `qvr check --target TARGET` does not capability-check an empty replacement module as success;
- the LSP publishes the same diagnostic at the offending expression;
- `qvr run` explains that the named program was omitted because of the recorded gap, rather than saying only that no entry exists; and
- direct transpilation reports the same stable code/message.

A source may still use an explicitly host-only form through an API that opts into the eager compiler, but portability must be an explicit mode, not silent fallback.

## Suggested implementation sequence

1. **Analysis foundation:** canonical visitor/rewriter; exhaustive tests; report `QiecModule.gap` in check/LSP/run.
2. **Fixed-shape map:** type/shape rules, inline and named lambda elaboration to `Comprehension`, `length`, reference and cross-target tests.
3. **Derived and general folds:** `logsumexp_over` via map/reduction; specify/fix eager fold order; add checked `Fold` if non-reduction folds are retained.
4. **Computation traversal:** grammar/AST/emitter, lowering to `Call`/`Bind`, stable sites, effect-row tests, multi-axis traversal.
5. **Builtin registry:** consolidate inventories and arities; classify/remove effectful pseudo-pure entries; decide checked broadcast semantics.
6. **Dynamic selection:** design packed or masked collections before enabling QIEC `filter`.
7. **Documentation and examples:** update the expression reference, QIEC reference, transpilation capability tables, and at least one nontrivial probabilistic tutorial/example using `map`, `traverse`, and reduction.

## Acceptance criteria

- The following checks, runs, serializes round-trip, and transpiles through the QIEC route:

  ```qvr
  define twice(xs : Tensor[Real]([3])) : Tensor[Real]([3]) !{} =
      return map(xs, x -> x * 2.0)
  ```

- Its lowered body is observationally equivalent to:

  ```qvr
  object K : FinSet 3

  define twice(xs : Tensor[Real]([3])) : Tensor[Real]([3]) !{} =
      return factor i : K in xs[i] * 2.0
  ```

- Mapping a tensor-valued body prepends the input's leading extent and preserves the body's shape.
- Captured values and shadowed names behave correctly for inline and let-bound lambdas.
- `length` has one documented checked type and works for every statically known leading extent.
- `fold` has one documented argument order; eager and QIEC execution agree.
- `logsumexp_over(xs, f)` agrees numerically with `logsumexp(map(xs, f))` on extreme inputs.
- A finite `traverse` can call an effect-free computation and a computation with a declared effect; trace/site identities are stable and distinct by iteration.
- The lexical-uncertainty RSA example can express the lexicon × world × utterance structure using finite map/traverse/reduction, with no hand-expanded list of speaker calls, and retains the expected posterior numerically.
- Every expression walker passes an exhaustiveness test over all current `LetExprNode` variants.
- A program containing an unsupported collection form cannot make `qvr check`, `qvr check --target`, or the LSP report clean success while being absent from `qvr run`.
- Every registered pure builtin has arity/type tests and either a QIEC lowering plus target capability path or an explicit host-only diagnostic.
- Grammar fixtures, canonical emission, parser snapshots, highlighting/semantic tokens for new reserved forms, reference docs, tutorials, and executable examples are updated together.

## Non-goals

- first-class runtime closures or unrestricted higher-order QIEC;
- arbitrary dynamic host-function or method dispatch in portable programs;
- a variable-length `filter` result before its type is specified; and
- silently adopting all PyTorch broadcasting and training-mode operations as source-language semantics.

## Related work

- #19 introduced tensor construction/`factor`, whose QIEC `Comprehension` is the right lowering target for fixed-shape map.
- #53 tracks per-target spelling for eager-path `LetExprCall` builtins. The registry work here should coordinate with it rather than duplicate renderer mappings.

## Relevant implementation sites

- `grammars/qvr/grammar.js`: shared let-expression grammar and unary lambda syntax
- `src/quivers/dsl/ast_nodes/let_expressions.py`: shared expression AST
- `src/quivers/dsl/compiler/programs.py`: eager closures, collection combinators, builtin dispatch, and factor execution
- `src/quivers/dsl/pure_builtins.py`: checked builtin inventory
- `src/quivers/dsl/qiec_lowering.py`: QIEC values, lambda rejection, macro expansion, reductions, rowwise operations, and factor-to-comprehension lowering
- `src/quivers/dsl/program_elaboration.py`: program gap handling and partial expression analyses
- `src/quivers/dsl/step_resolution.py`: a separate free-value walker
- `src/quivers/qiec/terms.py`, `checking.py`, `substitution.py`, `serialization.py`, and `evaluator.py`: first-order value terms
- `src/quivers/transpile/qiec_ir.py` and `src/quivers/transpile/renderers/_qiec.py`: checked IR and host-runtime emission
- `src/quivers/dsl/compiler/core.py` and `src/quivers/cli/check.py`: silent `QiecModule.gap` fallback
